shared-libs/cht-datasource/src/local/person.ts

```diff
@@ -82,4 +89,138 @@ export namespace v1 {
       )(limit, skip) as Page<Person.v1.Person>;
     };
   };
+
+  /** @internal */
+  export const create = ({
+    medicDb,
+    settings
+  }: LocalDataContext) => {
+    const createPersonDoc = createDoc(medicDb);
+    const getPersonDoc = getDocById(medicDb);
+    const ensureHasValidParentFieldAndReturnParentDoc = async (
+      input: Record<string, unknown>,
+      contactTypeObject: Record<string, unknown>
+    ): Promise<Nullable<Doc>> => {
+      const parentDoc = await getPersonDoc(input.parent as string);
+      if (parentDoc === null) {
+        throw new InvalidArgumentError(
+          `Parent with _id ${input.parent as string} does not exist.` //NoSONAR
+        );
+      }
+      // Check whether parent doc's `contact_type` or `type`(if `contact_type` is absent)
+      // matches with any of the allowed parents type.
+       
+      const typeToMatch = (parentDoc as Input.v1.PersonInput).contact_type 
+      || (parentDoc as Input.v1.PersonInput).type;
+      const parentTypeMatchWithAllowedParents = (contactTypeObject.parents as string[])
+        .find(parent => parent === typeToMatch);
+
+      if (!(parentTypeMatchWithAllowedParents)) {
+        throw new InvalidArgumentError(
+          `Parent of type ${JSON.stringify(typeToMatch)} is not allowed for ${JSON.stringify(
+            input.contact_type
+          )} type`
+        );
+      }
+      return parentDoc;
+    };
+
+    const validatePersonParent = async (
+      contactTypeObject: Record<string, unknown>,
+      input: Record<string, unknown>
+    ): Promise<Nullable<Doc>> => {
+      if (hasField(contactTypeObject, {name: 'parents', type: 'object'})) {
+        return await ensureHasValidParentFieldAndReturnParentDoc(input, contactTypeObject);
+      }
+
+      throw new InvalidArgumentError(
+        `Invalid type of person, cannot have parent for ${
+          JSON.stringify(input.contact_type)
+        } type`
+      );
+    };
+
+    const appendParent = async (
+      typeFoundInSettingsContactTypes: Record<string, unknown> | undefined,
+      input: Input.v1.PersonInput
+    ) => {
+      let parentDoc: Nullable<Doc> = null;
+      if (typeFoundInSettingsContactTypes) {
+        parentDoc = await validatePersonParent(typeFoundInSettingsContactTypes, input);
+      } else if (input.parent) {
+        parentDoc = await getPersonDoc(input.parent);
+      }
+
+      if (parentDoc === null) {
+        throw new InvalidArgumentError(
+          `Parent with _id ${input.parent} does not exist.`
+        );
+      }
+      input = addParentToInput(input, 'parent', parentDoc);
+      return input;
+    };
+
+    return async (input: Input.v1.PersonInput): Promise<Person.v1.Person> => {
+      input = Input.v1.validatePersonInput(input);
+      if (hasField(input, { name: '_rev', type: 'string', ensureTruthyValue: true })) {
+        throw new InvalidArgumentError('Cannot pass `_rev` when creating a person.');
+      }
```

Comment:
Can we do this in the `Input.v1.validatePersonInput` check?  (Maybe even down in the ContactInput logic...)